src/communication/kernels/all_reduce.cpp

```diff
@@ -0,0 +1,162 @@
+                           RequireContiguous::Yes>(std::forward<TileOutSender>(tile_out),
+                                                   std::move(all_reduce_final));
+}
+
+template <class T, Device D>
+auto allReduceInPlace(const Communicator& comm, MPI_Op reduce_op, const matrix::Tile<T, D>& tile,
+                      MPI_Request* req) {
+  DLAF_ASSERT(tile.is_contiguous(), "");
+
+  auto msg = comm::make_message(common::make_data(tile));
+  DLAF_MPI_CHECK_ERROR(
+      MPI_Iallreduce(MPI_IN_PLACE, msg.data(), msg.count(), msg.mpi_type(), reduce_op, comm, req));
+}
+
+DLAF_MAKE_CALLABLE_OBJECT(allReduceInPlace);
+}
+
+template <class T, Device D>
+[[nodiscard]] pika::execution::experimental::unique_any_sender<matrix::Tile<T, D>> scheduleAllReduce(
+    pika::execution::experimental::unique_any_sender<dlaf::common::PromiseGuard<Communicator>> pcomm,
+    MPI_Op reduce_op,
+    pika::execution::experimental::unique_any_sender<pika::shared_future<matrix::Tile<const T, D>>>
+        tile_in,
+    pika::execution::experimental::unique_any_sender<matrix::Tile<T, D>> tile_out) {
+  return internal::scheduleAllReduce(std::move(pcomm), reduce_op, std::move(tile_in),
+                                     std::move(tile_out));
+}
+
+// TODO: This is only for a test (test_collective_async)
+DLAF_SCHEDULE_ALL_REDUCE_SFTILE_ETI(, int, Device::CPU);
+
+template <class T, Device D>
+[[nodiscard]] pika::execution::experimental::unique_any_sender<matrix::Tile<T, D>> scheduleAllReduceInPlace(
+    pika::execution::experimental::unique_any_sender<dlaf::common::PromiseGuard<Communicator>> pcomm,
+    MPI_Op reduce_op, pika::execution::experimental::unique_any_sender<matrix::Tile<T, D>> tile) {
+  using dlaf::comm::CommunicationDevice_v;
+  using dlaf::comm::internal::allReduceInPlace_o;
+  using dlaf::comm::internal::transformMPI;
+  using dlaf::internal::CopyFromDestination;
+  using dlaf::internal::CopyToDestination;
+  using dlaf::internal::RequireContiguous;
+  using dlaf::internal::whenAllLift;
+  using dlaf::internal::withTemporaryTile;
+
+  auto all_reduce_in_place = [reduce_op, pcomm = std::move(pcomm)](auto const& tile_comm) mutable {
+    return whenAllLift(std::move(pcomm), reduce_op, std::cref(tile_comm)) |
+           transformMPI(allReduceInPlace_o);
+  };
+
+  // The tile has to be copied both to and from the temporary tile since the
+  // reduction is done in-place. The reduction is explicitly done on CPU memory
+  // so that we can manage potential asynchronous copies between CPU and GPU. A
+  // reduction requires contiguous memory.
+  return withTemporaryTile<CommunicationDevice_v<D>, CopyToDestination::Yes, CopyFromDestination::Yes,
+                           RequireContiguous::Yes>(std::move(tile), std::move(all_reduce_in_place));
+}
+
+// TODO: This is only for a test (test_collective_async)
+DLAF_SCHEDULE_ALL_REDUCE_IN_PLACE_ETI(, int, Device::CPU);
```

Comment:
This is the only open TODO from my side. This instantiation is only needed for one test which uses `int` as the element type. We can:
1. Keep it, since it's only one more element type (for a few of the communication kernels) and the cost is not _that_ high relatively speaking
2. Change the element type used in the test to be one of the floating point types that we anyway instantiate
3. Put the definitions of the kernels in the header files so that the test can instantiate the `int` case. This will keep the other instantiations in `dlaf.core` so it won't increase bloat but may increase compile times since the definitions are now in the header files.

My preference would be 2. Do you have a preference?

Reply:
I dont like much the idea of not supporting integer communication (in particular SizeType is needed if communications involves indices).

Reply:
See https://github.com/eth-cscs/DLA-Future/pull/674#issuecomment-1293771911.